# Task 1: Conceptual Question


Q. Do you agree with the following statement?

“Unsupervised learning is not suitable for classification problems, but supervised learning and reinforcement learning are.”

- Give reasons for your answer
- Justify your explanation clearly

A.

I partially disagree with the statement.

Unsupervised learning is generally not suitable for classification problems because it does not use labeled data. In classification tasks, the model needs input-output pairs (for example, email → spam or not spam). Since unsupervised learning does not have such labels, it cannot directly perform classification.
Supervised learning is suitable for classification because the model is trained using labeled data. It learns the relationship between input and correct output, and can then predict the class of new data accurately.

However, reinforcement learning is not typically used for classification problems. Although it uses rewards and penalties, it is mainly designed for sequential decision-making tasks where the correct action is not known in advance. In classification, the correct answer is already available, so supervised learning is more appropriate.

Therefore, supervised learning is suitable for classification, unsupervised learning is not directly suitable, and reinforcement learning is generally not used for this purpose.


# Task 2: Choosing the Right Metric


For each of the following machine learning tasks:

- Select a suitable evaluation metric
- Explain why this metric is appropriate


1. Spam Emails: Classify emails as spam or not spam
   - False negatives are acceptable
   - False positives are annoying

   Precision is the most suitable evaluation metric for spam email classification because false positives (i.e., misclassifying legitimate emails as spam) are highly undesirable. Precision ensures that when the model predicts an email as spam, it is highly likely to be correct, thereby improving user trust and system reliability.


2. Medical Diagnosis: Predict whether a patient has a rare disease
   - Missing a positive case is very serious
   - False alarms are less critical

   Recall is the most suitable evaluation metric for medical diagnosis of a rare disease because false negatives (i.e., failing to detect an actual disease case) can have severe or life-threatening consequences. Recall ensures that the model identifies as many true positive cases as possible, minimizing missed diagnoses.


3. Rental Prices: Estimate monthly rent based on features
   - Large errors are not much worse than multiple small errors

   Mean Absolute Error (MAE) is the most suitable metric for rental price prediction because it treats all errors equally and provides a clear interpretation of the average prediction error. Since large errors are not significantly more critical than small errors in this context, MAE offers a simple and reliable evaluation measure.


4. Weather Prediction: Predict tomorrow’s average temperature
   - Both small and large errors matter
   - Large errors should be penalized more

   MSE/RMSE is the most suitable evaluation metric for weather prediction because it penalizes large errors more heavily than small ones. This is important since large deviations in temperature prediction can lead to significant real-world consequences. RMSE is especially useful as it provides error in the same unit as the predicted variable, making interpretation easier


5. Credit Card Fraud: Detect fraudulent transactions
   - Fraud cases are rare
   - Need to detect them reliably
   - But too many false alarms are costly

   F1 Score is the most suitable evaluation metric for credit card fraud detection because both false positives and false negatives are important. It provides a balance between precision and recall, ensuring that the model effectively detects fraudulent transactions while minimizing incorrect blocking of legitimate transactions, especially in highly imbalanced datasets.


# Task 3: Training vs Testing


- Explain what a training dataset is
- Explain what a test dataset is
- Why should a machine learning model not be evaluated on the same data it was trained on?

  The training dataset is the data used to teach a machine learning model, where the model learns patterns and relationships between inputs and outputs by adjusting its internal parameters. The test dataset is a separate set of data that the model has never seen before, and it is used after training to evaluate how well the model performs on new, unseen data. A model should not be evaluated on the same data it was trained on because it may simply memorize the training data instead of learning general patterns, leading to overly high and misleading performance results; this problem is known as overfitting. As a result, although the model may perform perfectly on training data, it may fail when applied to real-world data, which is why a separate test dataset is necessary to measure the model’s true ability to generalize.


# Exercise 3: Machine Learning Basics

In this notebook, we will prepare data and build simple models to predict whether a passenger survived the Titanic disaster.


## Task 1: Loading the Dataset and Selecting Features

- Load the Titanic dataset into a pandas DataFrame.
- Research what the columns mean and argue which columns have no value for survival prediction: https://www.kaggle.com/datasets/yasserh/titanic-dataset/data.
- Choose 3 columns to use as features for the model. Include at least one numerical and one categorical feature.
- Extract each chosen column, as well as the target column (`Survived`), into a separate 1D NumPy array.

_Note: Make sure your data is clean!_


In [1]:
import pandas as pd
import numpy as np


df = pd.read_csv("../data/titanic.csv")

df.head()

# we are selecting sex, age and pclass as features becuase women, children and first class passengers were given priority during the evacuation.
df['Age'] = df['Age'].fillna(df['Age'].mean())
df = df.dropna(subset=['Sex', 'Pclass'])

# Features
age = df['Age'].to_numpy()
sex = df['Sex'].to_numpy()
pclass = df['Pclass'].to_numpy()

# Target    
y= df['Survived'].to_numpy()

# print(df[['Age', 'Sex', 'Pclass']].isnull().sum())


df.drop('Cabin',axis=1,inplace=True)
df = df.dropna(subset=['Embarked'])
print(df.isnull().sum())


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


## Task 2: Encoding / Scaling Attributes

- Apply **Min-Max Scaling** to the arrays with numerical attributes.
- Apply **One-Hot Encoding** to the arrays with categorical attributes. Since we long longer deal with columns, the encoding should result in a two-dimensional array. Check the example to better understand how the output should look.

_Note: Try to implement the function yourself, using only numpy and native python._


In [2]:
import numpy as np

# One-hot encoding example
example_array = np.array(["apple", "banana", "pineapple", "banana"])
example_array_oh = np.array([
    [1, 0, 0], # apple
    [0, 1, 0], # banana
    [0, 0, 1], # pineapple
    [0, 1, 0] # banana
])



# Min max Scalling for Age
age_min = np.min(age)
age_max = np.max(age)

age_scaled = (age-age_min)/(age_max-age_min)

print(age_scaled[:5])




# OneHot Encoding for Sex and PClass
unique_sex = np.unique(sex)

sex_oh = np.zeros((len(sex), len(unique_sex)))

for i, val in enumerate(sex):
    index = np.where(unique_sex == val)[0][0]
    sex_oh[i, index] = 1

print(sex_oh[:5])


unique_pclass = np.unique(pclass)

pclass_oh = np.zeros((len(pclass),len(unique_pclass)))

for i,val in enumerate(pclass):
    index = np.where(unique_pclass==val)[0][0]
    pclass_oh[i,index]=1

print(pclass_oh[:5])

[0.27117366 0.4722292  0.32143755 0.43453129 0.43453129]
[[0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]]
[[0. 0. 1.]
 [1. 0. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 0. 1.]]


## Task 3: Concatenating Features into a 2D Array

- Concatenate the numerical and one-hot encoded categorical arrays into a single 2D NumPy array.
- The resulting array should have one row per passenger and one column per feature

_Note: check the example to better understand the format._


In [3]:
import numpy as np

categories_oh = np.array([
    [1, 0, 0],  # apple
    [0, 1, 0],  # banana
    [0, 0, 1],  # pineapple
    [0, 1, 0]   # banana
])
sizes = np.array([22.0, 38.0, 26.0, 35.0])
concatenated = np.array([
    [1, 0, 0, 22.0],
    [0, 1, 0, 38.0],
    [0, 0, 1, 26.0],
    [0, 1, 0, 35.0]
])


age_scaled = age_scaled.reshape(-1,1)

X = np.concatenate((age_scaled,sex_oh,pclass_oh),axis=1)

print(X[:5])
print(X.shape)

[[0.27117366 0.         1.         0.         0.         1.        ]
 [0.4722292  1.         0.         1.         0.         0.        ]
 [0.32143755 1.         0.         0.         0.         1.        ]
 [0.43453129 1.         0.         1.         0.         0.        ]
 [0.43453129 0.         1.         0.         0.         1.        ]]
(891, 6)


## Task 4: Train-Test Split

Not all data should be used for training, some should be reserved for evaluating the model. Use a split of 70% of the data for training and 30% for testing.

- Split the feature data into `X_train` and `X_test`
- Split the target data into `y_train` and `y_test`

_Note: Try to implement the code yourself, using only numpy and native python._


In [4]:
indices = np.arange(len(X))
np.random.shuffle(indices)

X=X[indices]
y= y[indices]

split_index = int(0.8*len(X))

X_train = X[:split_index]
X_test = X[split_index:]

y_train = y[:split_index]
y_test = y[split_index:]

## Task 5: Decision Tree Model (Given Code)

Here we provide you with code to train a Decision Tree classifier on the same processed data.

- The method `fit` is used to train the decision tree on the training data (X_train, y_train).
- The method `predict` is used to generate predictions for unseen data (X_test).
- `preds_dt` contains the decision tree's predictions on the test data.

Do not worry if you do not understand all details yet. You just need to run the code and everything should work fine.

_Note: The package scikit-learn needs to be installed for the environment!_

_Second Note: If you run into an error, check if the arrays `X_train`, `y_train` and `X_test` are formatted correctly!_


In [5]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier()
print("Training the Decision Tree Classifier on X_train and y_train...")
dt.fit(X_train, y_train)

print("Making predictions on X_test...")
preds_dt = dt.predict(X_test)

print(f"First 10 predictions from the Decision Tree Classifier:\n{preds_dt[:10]}")
print(f"First 10 actual targets from y_test:\n{y_test[:10]}")

Training the Decision Tree Classifier on X_train and y_train...
Making predictions on X_test...
First 10 predictions from the Decision Tree Classifier:
[0 0 0 0 0 1 0 1 1 0]
First 10 actual targets from y_test:
[0 0 1 0 0 1 0 1 1 1]


## Task 6: Implementing a Guessing Model

Implement a simple model that ignores input features and only learns survival probabilities from the training data. It should have the functions `fit` and `predict`, that use the same parameters and serve a similar function as they did for the Decision Tree.

- The method `fit` should ignore the samples in `X_train` and simply learn the general probability of survival from the targets in `y_train`.
- The method `predict` should return array of predictions based on these learned probabilities for an input feature array.

Using your own model, call `fit` on `X_train` and `y_train` for training, then call `predict` on `X_test` to receive the predictions for the test data.


In [6]:
import numpy as np

class GuessingModel:
    def fit(self, X, y):
        self.survival_prob = np.mean(y)

    def predict(self, X):
        predictions = np.random.choice(
            [0, 1],
            size=len(X),
            p=[1 - self.survival_prob, self.survival_prob]
        )
        return predictions


        

In [7]:
gm = GuessingModel()

print("Training Guessing Model")
gm.fit(X_train, y_train)

print("Making predictions")
preds_gm = gm.predict(X_test)

print("First 10 predictions:")
print(preds_gm[:10])

Training Guessing Model
Making predictions
First 10 predictions:
[0 0 0 0 0 0 0 1 1 0]


## Task 7: Evaluation and Interpretation

- What kind of machine learning task is predicting survival of passengers?
- Suggest 2 evaluation metrics you could use for this task, and explain why.
- Compute both metrics for the guessing model and the decision tree.
- Interpret the results: how do the models compare?

_Note: Try to implement the code yourself, using only numpy and native python._


1. This was Binary Classification machine learning task because output has only 2 classes.
2. For predicting Titanic survival, two suitable evaluation metrics are Accuracy and F1 Score. Accuracy measures the overall percentage of correct predictions made by the model, so it is useful for understanding the model’s general performance. However, accuracy alone can sometimes be misleading if one class appears more often than the other. F1 Score is also important because it balances precision and recall, which helps measure how well the model identifies survivors while also avoiding too many incorrect predictions. Since the Titanic dataset is not perfectly balanced, using both Accuracy and F1 Score gives a better evaluation of the model’s performance.


In [8]:
# Accuracy
def accuracy(y_true, y_pred):
    correct = np.sum(y_true == y_pred)
    return correct / len(y_true)

# F1 Score
def f1_score(y_true, y_pred):
    tp = np.sum((y_true == 1) & (y_pred == 1))
    fp = np.sum((y_true == 0) & (y_pred == 1))
    fn = np.sum((y_true == 1) & (y_pred == 0))

    precision = tp / (tp + fp) if (tp + fp) != 0 else 0
    recall = tp / (tp + fn) if (tp + fn) != 0 else 0

    f1 = 2 * precision * recall / (precision + recall) if (precision + recall) != 0 else 0
    return f1

# Decision Tree
acc_dt = accuracy(y_test, preds_dt)
f1_dt = f1_score(y_test, preds_dt)

# Guessing Model
acc_gm = accuracy(y_test, preds_gm)
f1_gm = f1_score(y_test, preds_gm)

print("Decision Tree Accuracy:", acc_dt)
print("Decision Tree F1 Score:", f1_dt)

print("Guessing Model Accuracy:", acc_gm)
print("Guessing Model F1 Score:", f1_gm)

Decision Tree Accuracy: 0.7988826815642458
Decision Tree F1 Score: 0.7142857142857143
Guessing Model Accuracy: 0.5307262569832403
Guessing Model F1 Score: 0.4166666666666667


4. The Decision Tree achieved higher accuracy and F1 score than the Guessing Model, indicating that it learned meaningful patterns from passenger features. The Guessing Model serves as a baseline and performs worse because it ignores all input features and predicts randomly based on survival probability.


### Excercise 4 - Task 1: Multi-Class Classification Evaluation


🔹 Task 1: Multi-Class Classification Evaluation
A machine learning model classifies satellite images into four land-cover types:

- Forest (F)
- Urban (U)
- Water (W)
- Agriculture (A)
  The model was tested on 100 images.

📊 Prediction Results
Forest (Actual)

- 35 total
- 30 predicted as Forest - TP
- 2 predicted as Urban -
- 3 predicted as Agriculture

Urban (Actual)

- 20 total
- 15 predicted as Urban
- 2 predicted as Forest
- 2 predicted as Water
- 1 predicted as Agriculture

Water (Actual)

- 10 total
- 8 predicted as Water
- 2 predicted as Urban

Agriculture (Actual)

- 35 total
- 25 predicted as Agriculture
- 5 predicted as Forest
- 5 predicted as Urban

🔹 Part 1: Construct the Multi-Class Confusion Matrix
Create a 4 × 4 confusion matrix:

- Rows → Actual values
- Columns → Predicted values
  Classes order:
- Forest (F)
- Urban (U)
- Water (W)
- Agriculture (A)

🔹 Part 2: Binary Decomposition
Break the 4 × 4 confusion matrix into four separate 2 × 2 confusion matrices:
One for each class:

- Forest (F)
- Urban (U)
- Water (W)
- Agriculture (A)

In each 2 × 2 matrix:

- Focus class = 1 (Yes)
- All other classes combined = 0 (No)

🔹 Part 3: Calculate Performance Metrics
Calculate the following metrics for the entire model:

- Macro-averaged Precision
- Macro-averaged Recall
- Micro-averaged Precision
- Micro-averaged Recall
- Macro-averaged F1-Score

💻 Implementation Hint
You may calculate the values:

- Manually (to understand the logic)
OR
- Using Python

Suggested Libraries:

- numpy → for matrix manipulation
- sklearn.metrics → for verification
  Useful functions:
- confusion_matrix
- classification_report

📌 Note

- Understand the confusion matrix carefully
- Calculate TP, FP, FN, TN correctly for each class
- Then calculate macro and micro averages


In [9]:
import numpy as np
from sklearn.metrics import classification_report, precision_score, recall_score, f1_score

conf_matrix = np.array([
    [30,  2,  0,  3],
    [ 2, 15,  2,  1],
    [ 0,  2,  8,  0],
    [ 5,  5,  0, 25],
])


y_true=[]
y_pred=[]

for i in range(len(conf_matrix)):
    for j in range(len(conf_matrix[i])):
        count = conf_matrix[i][j]
        y_true.extend([i]*count)
        y_pred.extend([j]*count)

classes = ['Forest', 'Urban', 'Water', 'Agriculture']

print("Confusion Matrix:\n ",conf_matrix)

print("Classification Report:\n ",classification_report(y_true, y_pred, target_names=classes))

macro_precision = precision_score(y_true, y_pred, average='macro')
macro_recall = recall_score(y_true, y_pred, average='macro')
macro_f1 = f1_score(y_true, y_pred, average='macro')

micro_precision = precision_score(y_true, y_pred, average='micro')
micro_recall = recall_score(y_true, y_pred, average='micro')

print("Macro Precision:", round(macro_precision, 4))
print("Macro Recall:", round(macro_recall, 4))
print("Macro F1 Score:", round(macro_f1, 4))
print("Micro Precision:", round(micro_precision, 4))
print("Micro Recall:", round(micro_recall, 4))


Confusion Matrix:
  [[30  2  0  3]
 [ 2 15  2  1]
 [ 0  2  8  0]
 [ 5  5  0 25]]
Classification Report:
                precision    recall  f1-score   support

      Forest       0.81      0.86      0.83        35
       Urban       0.62      0.75      0.68        20
       Water       0.80      0.80      0.80        10
 Agriculture       0.86      0.71      0.78        35

    accuracy                           0.78       100
   macro avg       0.77      0.78      0.77       100
weighted avg       0.79      0.78      0.78       100

Macro Precision: 0.7745
Macro Recall: 0.7804
Macro F1 Score: 0.7741
Micro Precision: 0.78
Micro Recall: 0.78


# Task 1: Multi-Class Classification Evaluation (Manual Solution)

# Part 1: Construct the 4 × 4 Confusion Matrix

Rows = **Actual Class**

Columns = **Predicted Class**

| Actual \ Predicted  | Forest (F) | Urban (U) | Water (W) | Agriculture (A) |   Total |
| ------------------- | ---------: | --------: | --------: | --------------: | ------: |
| **Forest (F)**      |         30 |         2 |         0 |               3 |      35 |
| **Urban (U)**       |          2 |        15 |         2 |               1 |      20 |
| **Water (W)**       |          0 |         2 |         8 |               0 |      10 |
| **Agriculture (A)** |          5 |         5 |         0 |              25 |      35 |
| **Total**           |     **37** |    **24** |    **10** |          **29** | **100** |

---

# Part 2: Binary Decomposition

## (a) Forest (F)

### Step 1: Calculate TP

```text
TP = 30
```

### Step 2: Calculate FP

Predicted as Forest but actually not Forest

```text
FP = 2 + 0 + 5

FP = 7
```

### Step 3: Calculate FN

Actual Forest but predicted as another class

```text
FN = 2 + 0 + 3

FN = 5
```

### Step 4: Calculate TN

```text
TN = 100 − TP − FP − FN

TN = 100 − 30 − 7 − 5

TN = 58
```

### 2 × 2 Matrix

|                   | Predicted Forest | Predicted Not Forest |
| ----------------- | ---------------: | -------------------: |
| Actual Forest     |               30 |                    5 |
| Actual Not Forest |                7 |                   58 |

---

## (b) Urban (U)

### TP

```text
TP = 15
```

### FP

```text
FP = 2 + 2 + 5

FP = 9
```

### FN

```text
FN = 2 + 2 + 1

FN = 5
```

### TN

```text
TN = 100 − 15 − 9 − 5

TN = 71
```

### 2 × 2 Matrix

|                  | Predicted Urban | Predicted Not Urban |
| ---------------- | --------------: | ------------------: |
| Actual Urban     |              15 |                   5 |
| Actual Not Urban |               9 |                  71 |

---

## (c) Water (W)

### TP

```text
TP = 8
```

### FP

```text
FP = 0 + 2 + 0

FP = 2
```

### FN

```text
FN = 2
```

### TN

```text
TN = 100 − 8 − 2 − 2

TN = 88
```

### 2 × 2 Matrix

|                  | Predicted Water | Predicted Not Water |
| ---------------- | --------------: | ------------------: |
| Actual Water     |               8 |                   2 |
| Actual Not Water |               2 |                  88 |

---

## (d) Agriculture (A)

### TP

```text
TP = 25
```

### FP

```text
FP = 3 + 1 + 0

FP = 4
```

### FN

```text
FN = 5 + 5 + 0

FN = 10
```

### TN

```text
TN = 100 − 25 − 4 − 10

TN = 61
```

### 2 × 2 Matrix

|                        | Predicted Agriculture | Predicted Not Agriculture |
| ---------------------- | --------------------: | ------------------------: |
| Actual Agriculture     |                    25 |                        10 |
| Actual Not Agriculture |                     4 |                        61 |

---

# Part 3: Performance Metrics

---

# Step 1: Precision for Each Class

## Forest

```text
Precision(F)

= TP / (TP + FP)

= 30 / (30 + 7)

= 30 / 37

= 0.8108
```

---

## Urban

```text
Precision(U)

= 15 / (15 + 9)

= 15 / 24

= 0.6250
```

---

## Water

```text
Precision(W)

= 8 / (8 + 2)

= 8 / 10

= 0.8000
```

---

## Agriculture

```text
Precision(A)

= 25 / (25 + 4)

= 25 / 29

= 0.8621
```

---

# Step 2: Recall for Each Class

## Forest

```text
Recall(F)

= 30 / (30 + 5)

= 30 / 35

= 0.8571
```

---

## Urban

```text
Recall(U)

= 15 / (15 + 5)

= 15 / 20

= 0.7500
```

---

## Water

```text
Recall(W)

= 8 / (8 + 2)

= 8 / 10

= 0.8000
```

---

## Agriculture

```text
Recall(A)

= 25 / (25 + 10)

= 25 / 35

= 0.7143
```

---

# Step 3: F1-Score for Each Class

## Forest

```text
F1

= 2 × Precision × Recall
  ------------------------
   Precision + Recall

= 2 × 0.8108 × 0.8571
  -------------------
   0.8108 + 0.8571

= 0.8333
```

---

## Urban

```text
F1

= 2 × 0.625 × 0.75
  ----------------
   1.375

= 0.6818
```

---

## Water

```text
F1

= 2 × 0.8 × 0.8
  -------------
     1.6

= 0.8000
```

---

## Agriculture

```text
F1

= 2 × 0.8621 × 0.7143
  -------------------
   1.5764

= 0.7813
```

---

# Step 4: Macro-Averaged Precision

```text
Macro Precision

= (0.8108 + 0.6250 + 0.8000 + 0.8621)

  -----------------------------------

                 4

= 3.0979 / 4

= 0.7745
```

---

# Step 5: Macro-Averaged Recall

```text
Macro Recall

= (0.8571 + 0.7500 + 0.8000 + 0.7143)

  -----------------------------------

                 4

= 3.1214 / 4

= 0.7804
```

---

# Step 6: Macro-Averaged F1-Score

```text
Macro F1

= (0.8333 + 0.6818 + 0.8000 + 0.7813)

  -----------------------------------

                 4

= 3.0964 / 4

= 0.7741
```

---

# Step 7: Micro-Averaged Precision

Total True Positives

```text
TP

= 30 + 15 + 8 + 25

= 78
```

Total False Positives

```text
FP

= 7 + 9 + 2 + 4

= 22
```

```text
Micro Precision

= TP / (TP + FP)

= 78 / (78 + 22)

= 78 / 100

= 0.7800
```

---

# Step 8: Micro-Averaged Recall

Total False Negatives

```text
FN

= 5 + 5 + 2 + 10

= 22
```

```text
Micro Recall

= TP / (TP + FN)

= 78 / (78 + 22)

= 78 / 100

= 0.7800
```

---

# Final Answers

| Metric          |      Value |
| --------------- | ---------: |
| Macro Precision | **0.7745** |
| Macro Recall    | **0.7804** |
| Micro Precision | **0.7800** |
| Micro Recall    | **0.7800** |
| Macro F1-Score  | **0.7741** |


In [10]:
# Save processed data
np.save("../data/X_train.npy", X_train)
np.save("../data/X_test.npy",  X_test)
np.save("../data/y_train.npy", y_train)
np.save("../data/y_test.npy",  y_test)

print("Data saved successfully!")

Data saved successfully!
